src/what-unsafe-does.md

```diff
@@ -16,18 +16,41 @@ to your program. You definitely *should not* invoke Undefined Behavior.
 Unlike C, Undefined Behavior is pretty limited in scope in Rust. All the core
 language cares about is preventing the following things:
 
-* Dereferencing null, dangling, or unaligned pointers
-* Reading [uninitialized memory][]
+* Dereferencing (using the `*` operator on) dangling, or unaligned pointers, or
+  wide pointers with invalid metadata (see below)
 * Breaking the [pointer aliasing rules][]
-* Producing invalid primitive values:
-    * dangling/null references
-    * null `fn` pointers
+* Unwinding into another language
+* Causing a [data race][race]
+* Executing code compiled with target features that the current thread of execution does
+  not support (see [`target_feature`][])
+* Producing invalid primitive values (either alone or as a field of a compound
+  type such as `enum`/`struct`/array/tuple):
     * a `bool` that isn't 0 or 1
     * an undefined `enum` discriminant
+    * null `fn` pointers
     * a `char` outside the ranges [0x0, 0xD7FF] and [0xE000, 0x10FFFF]
-    * A non-utf8 `str`
-* Unwinding into another language
-* Causing a [data race][race]
+    * a `!` (all values are invalid for this type)
+    * dangling/unaligned references, references that do themselves point to
+      invalid values, or wide references (to a dynamically sized type) with
+      invalid metadata
+        * slice metadata is invalid if the slice has a total size larger than
+          `isize::MAX` bytes in memory
+        * `dyn Trait` metadata is invalid if it is not a pointer to a vtable for
+          `Trait` that matches the actual dynamic trait the reference points to
+    * a non-utf8 `str`
+    * an integer (`i*`/`u*`), floating point value (`f*`), or raw pointer read from
+      [uninitialized memory][]
+    * an invalid library type with custom invalid values, such as a `NonNull` or
+      the `NonZero` family of types, that is 0
+
+"Producing" a value happens any time a value is assigned, passed to a
+function/primitive operation or returned from a function/primitive operation.
```

Comment:
suggested reword and massive clarification:

Many have trouble accepting the consequences of invalid values, so they merit some extra discussion. The claim being made here is a very strong one, so read carefully.

A value is *produced* whenever it is assigned, passed to something, or returned from something. Keep in mind references get to assume their referents are valid, so you can't even create a reference to an invalid value. Additionally, uninitialized memory is **always invalid**, so you can't assign it to anything, pass it to anything, return it from anything, or take a reference to it. (Padding bytes are not technically part of a value's memory, and so may be left uninitialized.)

In simple and blunt terms: you cannot ever even **suggest** the existence of an invalid value. No, it's not ok if you "don't use" or "don't read" the value. Invalid values are **instant Undefined Behaviour**. The only correct way to manipulate memory that could be invalid is with raw pointers using methods like `write` and `copy`. If you want to leave a local variable or struct field uninitialized (or otherwise invalid), you *must* use a union or enum which clearly indicates at the type level that this memory may contain no values (see MaybeUninit for details).

Reply:
I applied most of your suggestions but this one is big enough that it is probably easier to hand the PR off to you. ;) I'd love to do a pass over what you got when you are done, if you don't mind.

I like this new text, as usual in you very pointed style! One comment though:

> Additionally, uninitialized memory is always invalid, so you can't assign it to anything

That's not true for `MaybeUninit`.